spirv_reflect.hpp

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright 2015-2018 ARM Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef SPIRV_CROSS_REFLECT_HPP
+#define SPIRV_CROSS_REFLECT_HPP
+
+#include "spirv_cross.hpp"
+#include "spirv_glsl.hpp"
+#include <utility>
+#include <vector>
+
+namespace spirv_cross
+{
+class CompilerReflection : public CompilerGLSL
```

Comment:
CompilerReflection should depend on Compiler, not CompilerGLSL.

Reply:
A lot of the code I depend on is in `CompilerGLSL`.  All the buffering code like `statement` are in that class.  I could probably extract that to either `Compiler` or some new intermediate class between `Compiler` and `CompilerGLSL` / `CompilerReflection`, but that would still leave a lot of other stuff out.  Currently, in order to ensure that all the internal structures like type member names have been fully populated I actually run the parent class method `CompilerGLSL::compile` before I start introspection.

Reply:
I'm looking into how much breaks if I change the base class.  The emitter stuff obviously, but I've actually been thinking of breaking out the simple JSON streaming stuff into it's own class so that if and when it can be replaced with something like `nlohmann::json` the work will be trivial.  The remaining things that break don't look as daunting to me as they initially did... some of them like `format_to_glsl` and `type_to_glsl` might be fixed by adding some static helpers that are used both by `CompilerGLSL` and `CompilerReflect`.  Others, like `buffer_is_packing_standard` are only used in limited scope where you suggest they might not be needed at all.  

I'll see what I can do.

Reply:
I forgot the other issue that popped up when I initially attempted to derive directly from `Compiler`.  The code in main.cpp uses `unique_ptr<CompilerGLSL> compiler;` to contain the instantiated compiler object, and then uses a significant number of methods (like `get_common_options`, `set_common_options`, `flatten_buffer_block`, etc) which are only visible on the `CompilerGLSL` class, not on the base class.  

I may be able to work around this by putting all the reflection work into the `else if (!args.reflect.empty())` block and doing an early return, but it might be ugly.  The alternative would seem to be a lot of new conditional blocks and static casting in the bulk of the main.cpp code.

Reply:
Keeping it as a subclass of GLSL is okay for now, don't spend a lot of time trying to rework this.

Reply:
Oops... too late... already did the refactoring.